File: tests/test_market_service.py
"""Tests for persistence, tick advancement and the query helpers."""

from __future__ import annotations

import pytest

from tests.conftest import GUILD


async def test_companies_are_seeded_on_first_use(market):
    engine = await market.get_engine(GUILD)
    assert len(engine.companies) == 20

    rows = await market.db.fetchall("SELECT symbol FROM companies")
    assert len(rows) == 20


async def test_seeding_is_idempotent(market, market_config):
    await market.get_engine(GUILD)
    # Simulate a fresh process against the same database.
    from hera.services.market import MarketService

    await market.db.execute("DELETE FROM companies")
    fresh = MarketService(market.db, market_config)
    await fresh.get_engine(GUILD)
    count = await market.db.fetchval("SELECT COUNT(*) FROM companies")
    assert count == 20


async def test_tick_persists_prices_and_history(market):
    result = await market.tick(GUILD)
    assert result.tick == 1

    rows = await market.db.fetchall("SELECT symbol, price FROM price_history WHERE tick = 1")
    assert len(rows) == 20

    state = await market.db.fetchone("SELECT tick, index_value FROM market_state WHERE guild_id = ?", (GUILD,))
    assert int(state["tick"]) == 1
    assert float(state["index_value"]) > 0


async def test_state_survives_a_restart(market, market_config):
    for _ in range(5):
        await market.tick(GUILD)
    company = await market.get_company(GUILD, "NOVA")
    price = company.price
    tick = (await market.snapshot(GUILD)).tick

    # Simulate a process restart with a brand new service over the same file.
    from hera.services.market import MarketService

    fresh = MarketService(market.db, market_config)
    await fresh.get_engine(GUILD)
    reloaded = await fresh.get_company(GUILD, "NOVA")
    assert reloaded.price == pytest.approx(price)
    assert (await fresh.snapshot(GUILD)).tick == tick


async def test_history_is_bounded_by_the_configured_limit(market):
    for _ in range(30):
        await market.tick(GUILD)
    rows = await market.db.fetchall(
        "SELECT COUNT(*) AS n FROM price_history WHERE symbol = 'NOVA'"
    )
    assert int(rows[0]["n"]) <= market.config.history_limit


async def test_history_returns_oldest_first(market):
    for _ in range(10):
        await market.tick(GUILD)
    history = await market.history("NOVA", limit=10)
    ticks = [tick for tick, _ in history]
    assert ticks == sorted(ticks)
    assert len(history) == 10


async def test_resolve_symbol_accepts_ticker_name_and_prefix(market):
    assert (await market.resolve_symbol(GUILD, "nova")).symbol == "NOVA"
    assert (await market.resolve_symbol(GUILD, "Novadyne Systems")).symbol == "NOVA"
    assert (await market.resolve_symbol(GUILD, "novadyne")).symbol == "NOVA"
    assert await market.resolve_symbol(GUILD, "zzzz") is None
    assert await market.resolve_symbol(GUILD, "") is None


async def test_news_is_recorded_and_readable(market):
    for _ in range(40):
        await market.tick(GUILD)
    items = await market.news(GUILD, limit=5)
    assert len(items) <= 5
    assert all(item.headline for item in items)


async def test_index_history_reconstructs_a_series(market):
    for _ in range(15):
        await market.tick(GUILD)
    series = await market.index_history(GUILD, limit=15)
    assert series
    ticks = [tick for tick, _ in series]
    assert ticks == sorted(ticks)
    assert all(value > 0 for _, value in series)


async def test_guilds_have_independent_markets(market):
    other = GUILD + 1
    for _ in range(3):
        await market.tick(GUILD)
    await market.tick(other)

    assert (await market.snapshot(GUILD)).tick == 3
    assert (await market.snapshot(other)).tick == 1


async def test_dividends_are_reported_by_the_tick(market):
    market.config = type(market.config)(dividend_tick_interval=3)
    seen = {}
    for _ in range(120):
        result = await market.tick(GUILD)
        seen.update(result.dividends)
    assert seen, "at least one company should have paid a dividend"

File: tests/test_market_service.ts
/** Tests for persistence, tick advancement and the query helpers. */
import { strict as assert } from 'node:assert';
import { test } from 'node:test';

import { MarketService } from '../hera/services/market';
import { Random } from '../hera/market/random';
import { makeHarness, testMarketConfig, GUILD } from './harness';

test('companies are seeded on first use', () => {
  const h = makeHarness();
  const engine = h.market.getEngine(GUILD);
  assert.equal(Object.keys(engine.companies).length, 20);
  assert.equal(h.db.fetchall('SELECT symbol FROM companies').length, 20);
  h.close();
});

test('seeding is idempotent', () => {
  const h = makeHarness();
  h.market.getEngine(GUILD);
  // Simulate a fresh process against the same database.
  h.db.execute('DELETE FROM companies');
  const fresh = new MarketService(h.db, testMarketConfig(), new Random(1));
  fresh.getEngine(GUILD);
  assert.equal(Number(h.db.fetchval('SELECT COUNT(*) AS n FROM companies')), 20);
  h.close();
});

test('a tick persists prices and history', () => {
  const h = makeHarness();
  const result = h.market.tick(GUILD);
  assert.equal(result.tick, 1);
  assert.equal(h.db.fetchall('SELECT symbol FROM price_history WHERE tick = 1').length, 20);
  const state = h.db.fetchone<{ tick: number; index_value: number }>(
    'SELECT tick, index_value FROM market_state WHERE guild_id = ?',
    [GUILD],
  );
  assert.equal(Number(state!.tick), 1);
  assert.ok(Number(state!.index_value) > 0);
  h.close();
});

test('state survives a restart', () => {
  const h = makeHarness();
  for (let i = 0; i < 5; i += 1) h.market.tick(GUILD);
  const price = h.market.getCompany(GUILD, 'NOVA')!.price;
  const tick = h.market.snapshot(GUILD).tick;

  // Simulate a process restart with a brand new service over the same file.
  const fresh = new MarketService(h.db, testMarketConfig(), new Random(1));
  fresh.getEngine(GUILD);
  assert.ok(Math.abs(fresh.getCompany(GUILD, 'NOVA')!.price - price) < 1e-9);
  assert.equal(fresh.snapshot(GUILD).tick, tick);
  h.close();
});

test('history is bounded by the configured limit', () => {
  const h = makeHarness();
  for (let i = 0; i < 80; i += 1) h.market.tick(GUILD);
  const count = Number(
    h.db.fetchval("SELECT COUNT(*) AS n FROM price_history WHERE symbol = 'NOVA'"),
  );
  // The prune keeps everything at or after ``tick - limit``, so the store can
  // hold one entry more than the nominal limit.
  assert.ok(count <= h.market.config.historyLimit + 1, `kept ${count} rows`);
  h.close();
});

test('history returns oldest first', () => {
  const h = makeHarness();
  for (let i = 0; i < 10; i += 1) h.market.tick(GUILD);
  const history = h.market.history('NOVA', 10);
  const ticks = history.map(([tick]) => tick);
  assert.deepEqual(ticks, [...ticks].sort((a, b) => a - b));
  assert.equal(history.length, 10);
  h.close();
});

test('resolveSymbol accepts ticker, name and prefix', () => {
  const h = makeHarness();
  assert.equal(h.market.resolveSymbol(GUILD, 'nova')!.symbol, 'NOVA');
  assert.equal(h.market.resolveSymbol(GUILD, 'Novadyne Systems')!.symbol, 'NOVA');
  assert.equal(h.market.resolveSymbol(GUILD, 'novadyne')!.symbol, 'NOVA');
  assert.equal(h.market.resolveSymbol(GUILD, 'zzzz'), undefined);
  assert.equal(h.market.resolveSymbol(GUILD, ''), undefined);
  h.close();
});

test('news is recorded and readable', () => {
  const h = makeHarness();
  for (let i = 0; i < 40; i += 1) h.market.tick(GUILD);
  const items = h.market.news(GUILD, 5);
  assert.ok(items.length <= 5);
  assert.ok(items.every((item) => item.headline));
  h.close();
});

test('index history reconstructs a series', () => {
  const h = makeHarness();
  for (let i = 0; i < 15; i += 1) h.market.tick(GUILD);
  const series = h.market.indexHistory(GUILD, 15);
  assert.ok(series.length > 0);
  const ticks = series.map(([tick]) => tick);
  assert.deepEqual(ticks, [...ticks].sort((a, b) => a - b));
  assert.ok(series.every(([, value]) => value > 0));
  h.close();
});

test('guilds have independent markets', () => {
  const h = makeHarness();
  const other = String(Number(GUILD) + 1);
  for (let i = 0; i < 3; i += 1) h.market.tick(GUILD);
  h.market.tick(other);
  assert.equal(h.market.snapshot(GUILD).tick, 3);
  assert.equal(h.market.snapshot(other).tick, 1);
  h.close();
});

test('dividends are reported by the tick', () => {
  const h = makeHarness();
  // A short dividend interval makes a payout certain within the window.
  const market = new MarketService(h.db, testMarketConfig({ dividendTickInterval: 3 }), new Random(1));
  const seen: Record<string, number> = {};
  for (let i = 0; i < 120; i += 1) {
    const result = market.tick(GUILD);
    Object.assign(seen, result.dividends);
  }
  assert.ok(Object.keys(seen).length > 0, 'at least one company should have paid a dividend');
  h.close();
});
